src/nupic/experimental/ExtendedTemporalMemory.cpp

```diff
@@ -0,0 +1,1316 @@
+/* ---------------------------------------------------------------------
+ * Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2013-2016, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ----------------------------------------------------------------------
+ */
+
+/** @file
+ * Implementation of ExtendedTemporalMemory
+ */
+
+#include <cstring>
+#include <climits>
+#include <iostream>
+#include <string>
+#include <iterator>
+#include <vector>
+
+#include <capnp/message.h>
+#include <capnp/serialize.h>
+#include <kj/std/iostream.h>
+
+#include <nupic/algorithms/Connections.hpp>
+#include <nupic/experimental/ExtendedTemporalMemory.hpp>
+
+using namespace std;
+using namespace nupic;
+using namespace nupic::algorithms::connections;
+using namespace nupic::experimental::extended_temporal_memory;
+
+#define EPSILON 0.000001
+
+ExtendedTemporalMemory::ExtendedTemporalMemory()
+{
+  version_ = 2;
```

Comment:
Why is this version 2? And I'd recommend just having a constant with the current version. The only place an instance with an older version will exist is in a serialized state and when you deserialize it you update it right away to the latest version. So you have to store the version when serializing but no need to store it as a member variable on the class when it can just be a constant.